File: app/src/main/java/jp/ac/u_tokyo/t/seo/station/app/KalmanFilter.java
package jp.ac.u_tokyo.t.seo.station.app;

import org.apache.commons.math3.linear.MatrixUtils;
import org.apache.commons.math3.linear.RealMatrix;

/**
 * @author Seo-4d696b75
 * @version 2020/06/28.
 */
class KalmanFilter {


    /**
     https://ja.wikipedia.org/wiki/%E8%B5%B7%E5%8B%95%E5%8A%A0%E9%80%9F%E5%BA%A6
     電車の起動加速度はせいぜい5.0km/h/s
     標準正規分布の累積確率分布関数で97.5%点は1.97
     標準偏差2.5km/h/sぐらい ~ 1.0m/s^2 のオーダー
     */
    private final double SIGMA = 1.0;

    static class Sample implements Cloneable {

        /**
         * 初期化した状態オブジェクトを返す
         * @param time 時刻
         * @param pos 観測現在位置
         * @param speed 観測現在速度, 観測値欠損の場合は0（適当な値）
         * @param variance 位置・速度の観測値の平均分散>=0、値が大きいほど観測精度が低いと解釈される
         * @return
         */
        static Sample initialize(long time, double pos, double speed, double variance){
            variance = Math.max(variance, 0);
            Sample state = new Sample(time, pos, speed, 0);

            state.state = MatrixUtils.createRealMatrix(new double[][]{{pos}, {speed}, {0}});
            state.p = MatrixUtils.createRealMatrix(new double[][]{{variance,0,0},{0,variance,0},{0,0,variance}});
            return state;
        }


        private Sample(long time, double pos, double speed, double acceleration){
            this.time = time;
            this.pos = pos;
            this.speed = speed;
            this.acceleration = acceleration;
        }

        final long time;
        final double pos, speed, acceleration;
        private RealMatrix state, p;

        public Sample clone() throws CloneNotSupportedException {
            Sample other = (Sample)super.clone();
            other.state = state.copy();
            other.p = p.copy();
            return other;
        }

    }

    KalmanFilter(){
        mH = MatrixUtils.createRealMatrix(new double[][]{{1,0,0}});
    }

    private final RealMatrix mH;

    Sample update(double pos, double err, long time, Sample latest){
        double delta = (time - latest.time) / 1000.0;
        RealMatrix f = MatrixUtils.createRealMatrix(new double[][]{{1,delta,delta*delta/2},{0,1,delta},{0,0,1}});
        RealMatrix g = MatrixUtils.createRealMatrix(new double[][]{{delta*delta*delta/6},{delta*delta/2},{delta}});
        //estimate
        RealMatrix estimatedState = f.multiply(latest.state);
        RealMatrix p = f.multiply(latest.p).multiply(f.transpose()).add(g.multiply(g.transpose()).scalarMultiply(SIGMA*SIGMA));
        //update
        double e = pos - mH.multiply(estimatedState).getEntry(0,0);
        double s = err*err + mH.multiply(p).multiply(mH.transpose()).getEntry(0,0);
        RealMatrix k = p.multiply(mH.transpose()).scalarMultiply(1.0 / s);
        RealMatrix nextState = estimatedState.add(k.scalarMultiply(e));
        RealMatrix nextP = MatrixUtils.createRealIdentityMatrix(3).subtract(k.multiply(mH)).multiply(p);
        Sample next = new Sample(
                time,
                nextState.getEntry(0,0),
                nextState.getEntry(1,0),
                nextState.getEntry(2,0)
        );
        next.state = nextState;
        next.p = nextP;
        return next;
    }


}

File: app/src/main/java/jp/seo/station/ekisagasu/polyline/KalmanFilter.kt
package jp.seo.station.ekisagasu.polyline

import org.apache.commons.math3.linear.MatrixUtils
import org.apache.commons.math3.linear.RealMatrix

/**
 * @author Seo-4d696b75
 * @version 2020/06/28.
 */
class KalmanFilter {

    companion object {

        /**
         * https://ja.wikipedia.org/wiki/%E8%B5%B7%E5%8B%95%E5%8A%A0%E9%80%9F%E5%BA%A6
         * 電車の起動加速度はせいぜい5.0km/h/s
         * 標準正規分布の累積確率分布関数で97.5%点は1.97
         * 標準偏差2.5km/h/sぐらい ~ 1.0m/s^2 のオーダー
         */
        const val SIGMA = 1.0
    }

    class Sample(
        val time: Long,
        val pos: Double,
        val speed: Double,
        val acceleration: Double,
        var state: RealMatrix,
        var p: RealMatrix
    ) : Cloneable {

        @Throws(CloneNotSupportedException::class)
        public override fun clone(): Sample {
            val other = super.clone() as Sample
            other.state = state.copy()
            other.p = p.copy()
            return other
        }

        companion object {
            /**
             * 初期化した状態オブジェクトを返す
             * @param time 時刻
             * @param pos 観測現在位置
             * @param speed 観測現在速度, 観測値欠損の場合は0（適当な値）
             * @param variance 位置・速度の観測値の平均分散>=0、値が大きいほど観測精度が低いと解釈される
             * @return
             */
            fun initialize(time: Long, pos: Double, speed: Double, variance: Double): Sample {
                val v = variance.coerceAtLeast(0.0)
                val state = MatrixUtils.createRealMatrix(
                    arrayOf(
                        doubleArrayOf(pos),
                        doubleArrayOf(speed),
                        doubleArrayOf(0.0)
                    )
                )
                val p = MatrixUtils.createRealMatrix(
                    arrayOf(
                        doubleArrayOf(v, 0.0, 0.0),
                        doubleArrayOf(0.0, v, 0.0),
                        doubleArrayOf(0.0, 0.0, v)
                    )
                )
                return Sample(time, pos, speed, 0.0, state, p)
            }
        }
    }

    private val mH: RealMatrix = MatrixUtils.createRealMatrix(arrayOf(doubleArrayOf(1.0, 0.0, 0.0)))

    fun update(pos: Double, err: Double, time: Long, latest: Sample): Sample {
        val delta = (time - latest.time) / 1000.0
        val f = MatrixUtils.createRealMatrix(
            arrayOf(
                doubleArrayOf(1.0, delta, delta * delta / 2),
                doubleArrayOf(0.0, 1.0, delta),
                doubleArrayOf(0.0, 0.0, 1.0)
            )
        )
        val g = MatrixUtils.createRealMatrix(
            arrayOf(
                doubleArrayOf(delta * delta * delta / 6),
                doubleArrayOf(delta * delta / 2),
                doubleArrayOf(delta)
            )
        )
        // estimate
        val estimatedState = f.multiply(latest.state)
        val p = f.multiply(latest.p).multiply(f.transpose())
            .add(g.multiply(g.transpose()).scalarMultiply(SIGMA * SIGMA))
        // update
        val e = pos - mH.multiply(estimatedState).getEntry(0, 0)
        val s = err * err + mH.multiply(p).multiply(mH.transpose()).getEntry(0, 0)
        val k = p.multiply(mH.transpose()).scalarMultiply(1.0 / s)
        val nextState = estimatedState.add(k.scalarMultiply(e))
        val nextP = MatrixUtils.createRealIdentityMatrix(3).subtract(k.multiply(mH)).multiply(p)
        return Sample(
            time,
            nextState.getEntry(0, 0),
            nextState.getEntry(1, 0),
            nextState.getEntry(2, 0),
            nextState,
            nextP
        )
    }
}
